source/common/config/datasource.cc

```diff
@@ -61,12 +43,109 @@ absl::StatusOr<std::string> read(const envoy::config::core::v3::DataSource& sour
+      return absl::InvalidArgumentError(
+          fmt::format("file {} size is {} bytes; maximum is {}", path, size, max_size));
+    }
+  }
+
+  auto file_content_or_error = file_system.fileReadToEnd(path);
+  RETURN_IF_STATUS_NOT_OK(file_content_or_error);
+
+  if (!allow_empty && file_content_or_error.value().empty()) {
+    return absl::InvalidArgumentError(fmt::format("file {} is empty", path));
+  }
+
+  return file_content_or_error.value();
+}
+
 absl::optional<std::string> getPath(const envoy::config::core::v3::DataSource& source) {
   return source.specifier_case() == envoy::config::core::v3::DataSource::SpecifierCase::kFilename
              ? absl::make_optional(source.filename())
              : absl::nullopt;
 }
 
+DynamicData::DynamicData(Event::Dispatcher& main_dispatcher,
+                         ThreadLocal::TypedSlotPtr<ThreadLocalData> slot,
+                         Filesystem::WatcherPtr watcher)
+    : dispatcher_(main_dispatcher), slot_(std::move(slot)), watcher_(std::move(watcher)) {}
+
+DynamicData::~DynamicData() {
+  if (!dispatcher_.isThreadSafe()) {
+    dispatcher_.post([to_delete = std::move(slot_)] {});
+  }
+}
+
+absl::string_view DynamicData::data() const {
+  const auto thread_local_data = slot_->get();
+  return thread_local_data.has_value() ? *thread_local_data->data_ : EMPTY_STRING;
+}
+
+absl::string_view DataSourceProvider::data() const {
+  if (absl::holds_alternative<std::string>(data_)) {
+    return absl::get<std::string>(data_);
+  }
+  return absl::get<DynamicData>(data_).data();
+}
+
+absl::StatusOr<DataSourceProvider>
+DataSourceProvider::create(const ProtoDataSource& source, OptRef<const ProtoWatchedDirectory> watch,
+                           Event::Dispatcher& main_dispatcher, ThreadLocal::SlotAllocator& tls,
+                           Api::Api& api, bool allow_empty, uint64_t max_size) {
+  auto initial_data_or_error = read(source, allow_empty, api, max_size);
+  RETURN_IF_STATUS_NOT_OK(initial_data_or_error);
+
+  if (!watch || source.specifier_case() != envoy::config::core::v3::DataSource::kFilename) {
+    return DataSourceProvider(std::move(initial_data_or_error).value());
+  }
+
+  auto slot = ThreadLocal::TypedSlot<DynamicData::ThreadLocalData>::makeUnique(tls);
+  slot->set([initial_data = std::make_shared<std::string>(
+                 std::move(initial_data_or_error.value()))](Event::Dispatcher&) {
+    return std::make_shared<DynamicData::ThreadLocalData>(initial_data);
+  });
+
+  const auto& filename = source.filename();
+  auto watcher = main_dispatcher.createFilesystemWatcher();
+  // DynamicData will ensure that the watcher is destroyed before the slot is destroyed.
+  // TODO(wbpcode): use Config::WatchedDirectory instead of directly creating a watcher
+  // if the Config::WatchedDirectory is exception-free in the future.
+  auto watcher_status = watcher->addWatch(
+      absl::StrCat(watch->path(), "/"), Filesystem::Watcher::Events::MovedTo,
```

Comment:
This is a very specific kind of watch, for MovedTo. It also relies on the directory mostly not having other stuff going on (or we see lots of updates unnecessarily). Is tihs specific set of choices over indexing on one use case you have?

Reply:
It's how you deal with a symlink swap in the other cases we watch on. It's not the only way to watch a file, and if you're dealing with a single file, you might want to also deal with cases like it being modified in-place. Otherwise, you need to document somewhere and give guidance the way to trigger an update is via a move.

Reply:
>  if you're dealing with a single file, you might want to also deal with cases like it being modified in-place.

I have no plan to do that for now, it will make thing much complex. Basically, this PR is only a tool function changes.
And we should document this limitation (only move is supported) at the new API when we actually use it in some scenario. (it should be #32789 as far as I know.)

Reply:
If the main use case of this is https://github.com/envoyproxy/envoy/pull/32789, what does the xDS API look like there? The PR under review is introducing the need to specify a watched directory, how will #32789 supply this?

Reply:
A new watch directory should be added to the DirectResponseAction. And if it's set explicitly, the watching will be enabled.

```
message DirectResponseAction {
  option (udpa.annotations.versioning).previous_message_type =
      "envoy.api.v2.route.DirectResponseAction";

  // Specifies the HTTP response status to be returned.
  uint32 status = 1 [(validate.rules).uint32 = {lt: 600 gte: 200}];

  // Specifies the content of the response body. If this setting is omitted,
  // no body is included in the generated response.
  //
  // .. note::
  //
  //   Headers can be specified using ``response_headers_to_add`` in the enclosing
  //   :ref:`envoy_v3_api_msg_config.route.v3.Route`, :ref:`envoy_v3_api_msg_config.route.v3.RouteConfiguration` or
  //   :ref:`envoy_v3_api_msg_config.route.v3.VirtualHost`.
  core.v3.DataSource body = 2;

  // Enable file watching if this field is set explicitly and ``filename`` is used in the ``body``. Only move action on the watched directory will result in a reloading.
  core.v3.WatchDirectory watch = 3;
}

```

Reply:
An API like this takes us back to @alyssawilk original question I think. Shouldn't this be something we explicitly select into if we need to provide a `WatchedDirectory` anyway?

Reply:
I guess you mean why we don't add a new specific config to do this directly and needn't to combine the `DataSource` and `WatchedDirecty` ? Like the `PathConfigSource` that could watch single file directly?

From my perspective, it's because I want to reuse the exist `DataSource` API. And make the watching as optional extension to the `DataSource` . If we create a new specific/independent API to complete this feature, then the new API would be a replacement of `DataSource`.

Another reason is the existing API's affection. The `TlsCertificate` use similar way to combine the `DataSource` and `WatchedDirectory`.